### DISCUSSION TITLE
Can't get it to run on Apple silicon.

### DISCUSSION THREAD

**Discussion by elchupacabrinski** [Mikubill/sd-webui-controlnet#860] [Q&A]
I always get the following error:
RuntimeError: slow_conv2d_forward_mps: input(device='cpu') and weight(device=mps:0')  must be on the same device

How can stuff even be on different devices without a GPU? :D

**Comment by baicuya** [comment]
I got same issue:RuntimeError: slow_conv2d_forward_mps: input(device='cpu') and weight(device=mps:0')  must be on the same device

**Reply by baicuya** [reply]
I have solved it ,you can add code /stable-diffusion-webui/extensions/sd-webui-controlnet/annotator/openpose/body.py  line48:
            if torch.backends.mps.is_available():
                mps_device = torch.device("mps:0")
                data = data.to(mps_device)

**Reply by sonluong1987** [reply]
Thank you!
If you use openpose full then add the above code to these lines as well:

hand.py line 54

        with torch.no_grad():  
            if torch.backends.mps.is_available():  
                mps_device = torch.device("mps:0")  
                data = data.to(mps_device)  
            output = self.model(data).cpu().numpy()  


face.py line 341

      with torch.no_grad():
          if torch.backends.mps.is_available():
              mps_device = torch.device("mps:0")
              x_data = x_data.to(mps_device)
          hs = self.model(x_data[None, ...])
          heatmaps = F.interpolate(
              hs[-1],
              (H, W),
              mode='bilinear', align_corners=True).cpu().numpy()[0]

**Comment by pandora523** [comment]
mine can't update media-pipe
I did install open-cv but it was uptodate

Python 3.10.9 (main, Jan 11 2023, 09:18:18) [Clang 14.0.6 ]
Commit hash: 22bcc7be428c94e9408f589966c2040187245d81
Installing requirements for Web UI
Installing sd-webui-controlnet requirement: mediapipe==0.9.1.0
Couldn't install sd-webui-controlnet requirement: mediapipe==0.9.1.0.
Command: "/Users/rei/stable-diffusion-webui/venv-torch-nightly/bin/python3.10" -m pip install mediapipe==0.9.1.0 --prefer-binary
Error code: 1
stdout: <empty>
stderr: ERROR: Could not find a version that satisfies the requirement mediapipe==0.9.1.0 (from versions: 0.9.3.0)
ERROR: No matching distribution found for mediapipe==0.9.1.0

[notice] A new release of pip available: 22.3.1 -> 23.1
[notice] To update, run: pip install --upgrade pip


Warning: Failed to install mediapipe==0.9.1.0, some preprocessors may not work.
Installing sd-webui-controlnet requirement: opencv-python

**Reply by pswai** [reply]
@pandora523 I tried modifying mediapipe to 0.9.3.0 in `extensions/sd-webui-controlnet/requirements.txt`, seems to work

**Reply by pandora523** [reply]
ohh it seems work thanks!

 the next problem I have is the annotator not working and having the same problem as above 😅
so do I just open .py  files and open it in text editor and then paste the code on the bottom of the code?

**Reply by pswai** [reply]
@pandora523 you mean the `slow_conv2d_forward_mps` issue above? Yes just use your text editor and paste @sonluong1987's code at the line numbers mentioned. If you are not familiar with Python, pay attention to the whitespaces, then you are good to go

**Reply by pandora523** [reply]
yes!😅 my stupidity is eating me alive  idid try the body.py in line 48 did I paste it in the right place?
sorry for keep bothering you 🥲
<img width="684" alt="scrensht" src="https://user-images.githubusercontent.com/131154780/232936683-31e89ac7-a8a4-4479-86d5-c9e6044e8df6.png">

**Reply by pswai** [reply]
> yes!😅 my stupidity is eating me alive idid try the body.py in line 48 did I paste it in the right place? sorry for keep bothering you 🥲 <img alt="scrensht" width="684" src="https://user-images.githubusercontent.com/131154780/232936683-31e89ac7-a8a4-4479-86d5-c9e6044e8df6.png">

No worries! You are almost there. In your case, it's the indentation that is incorrect. Let me take a screenshot of mine:
<img width="591" alt="image" src="https://user-images.githubusercontent.com/5335123/232937464-bd269f75-29a8-4182-ba22-a71385d507d6.png">

Then you can do the same for `hands.py` and `face.py`. Let's try this and see if it works.

Side note: Completely optional, but I suggest you to use a code editor like VSCode that can make your life easier in editing 😄

**Reply by pandora523** [reply]
im not sure if I did it right 
<img width="1280" alt="Screenshot 2023-04-19 at 10 12 23" src="https://user-images.githubusercontent.com/131154780/232941629-4c19fcc5-5027-471f-9b2b-d8cfd3bc743d.png">

<img width="1280" alt="Screenshot 2023-04-19 at 10 12 12" src="https://user-images.githubusercontent.com/131154780/232941658-dd870012-11b4-4e46-a846-4b13a383d15a.png">

<img width="1280" alt="Screenshot 2023-04-19 at 10 12 16" src="https://user-images.githubusercontent.com/131154780/232941688-11dcf444-f37a-43ed-b437-b397728c7834.png">


im now getting this on Terminal.      ModuleNotFoundError: No module named 'dumpy'

**Reply by pswai** [reply]
> im not sure if I did it right <img alt="Screenshot 2023-04-19 at 10 12 23" width="1280" src="https://user-images.githubusercontent.com/131154780/232941629-4c19fcc5-5027-471f-9b2b-d8cfd3bc743d.png">
> 
> <img alt="Screenshot 2023-04-19 at 10 12 12" width="1280" src="https://user-images.githubusercontent.com/131154780/232941658-dd870012-11b4-4e46-a846-4b13a383d15a.png">
> 
> <img alt="Screenshot 2023-04-19 at 10 12 16" width="1280" src="https://user-images.githubusercontent.com/131154780/232941688-11dcf444-f37a-43ed-b437-b397728c7834.png">
> 
> im now getting this on Terminal. ModuleNotFoundError: No module named 'dumpy'

Your `body.py` is correct, but not the other two files.

For `hand.py`, the problem is the indentation (number of whitespaces). You need to keep the `if` 4 spaces away from the `with`, telling Python that that part of the code is "inside" the `with`. The same thing with line 55 and 56. Here's a screenshot of mine.

<img width="485" alt="image" src="https://user-images.githubusercontent.com/5335123/232942474-840e2d58-a0e6-42e3-a3b6-d82e4e456831.png">

Similar for `face.py`:

<img width="589" alt="image" src="https://user-images.githubusercontent.com/5335123/232942607-4cda9367-206e-4d58-8035-8b3b6c2d6e5a.png">

Note that the code being pasted to all 3 files is the same:
```python
if torch.backends.mps.is_available():
    mps_device = torch.device("mps:0")
    x_data = x_data.to(mps_device)
```
Just need to get the spacing right!

**Reply by pandora523** [reply]
I think it works! 

About this one idid rename the the line 2 from dumpy to numpy. and have no idea what im doing just saw the code saying numpy so I did renamed it and it works!? lol. im not sure if this solves all my problems but that was really helpful !😆 
God Blessed You By The Computer Gods!!!

File "/Users/lei/stable-diffusion-webui/extensions/sd-webui-controlnet/annotator/openpose/body.py", line 2, in <module>
    import dumpy as np
ModuleNotFoundError: No module named 'dumpy'

**Reply by pswai** [reply]
> I think it works!
> 
> About this one idid rename the the line 2 from dumpy to numpy. and have no idea what im doing just saw the code saying numpy so I did renamed it and it works!? lol. im not sure if this solves all my problems but that was really helpful !😆 God Blessed You By The Computer Gods!!!
> 
> File "/Users/lei/stable-diffusion-webui/extensions/sd-webui-controlnet/annotator/openpose/body.py", line 2, in import dumpy as np ModuleNotFoundError: No module named 'dumpy'

Glad that it works! For dumpy/numpy it's very weird. For my file it's already `numpy`. But great that you figured it out! Now have fun!

**Reply by shelly-luo** [reply]
I got more errors after following this step😭
Startup time: 13.6s (import torch: 0.8s, import gradio: 3.7s, import ldm: 0.3s, other imports: 0.6s, list extensions: 0.3s, load scripts: 0.8s, load SD checkpoint: 6.5s, create ui: 0.2s, gradio launch: 0.3s).
Traceback (most recent call last):
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/gradio/routes.py", line 337, in run_predict
    output = await app.get_blocks().process_api(
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/gradio/blocks.py", line 1015, in process_api
    result = await self.call_function(
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/gradio/blocks.py", line 833, in call_function
    prediction = await anyio.to_thread.run_sync(
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/Users/shelly/Documents/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 373, in run_annotator
    result, is_image = preprocessor(img, res=pres, thr_a=pthr_a, thr_b=pthr_b)
  File "/Users/shelly/Documents/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/processor.py", line 138, in openpose
    result = model_openpose(img)
  File "/Users/shelly/Documents/stable-diffusion-webui/extensions/sd-webui-controlnet/annotator/openpose/__init__.py", line 84, in __call__
    candidate, subset = self.body_estimation(oriImg)
  File "/Users/shelly/Documents/stable-diffusion-webui/extensions/sd-webui-controlnet/annotator/openpose/body.py", line 48, in __call__
    Mconv7_stage6_L1, Mconv7_stage6_L2 = self.model(data)
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/Users/shelly/Documents/stable-diffusion-webui/extensions/sd-webui-controlnet/annotator/openpose/model.py", line 116, in forward
    out1 = self.model0(x)
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/container.py", line 217, in forward
    input = module(input)
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/Users/shelly/Documents/stable-diffusion-webui/extensions-builtin/Lora/lora.py", line 202, in lora_Conv2d_forward
    return lora_forward(self, input, torch.nn.Conv2d_forward_before_lora(self, input))
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 463, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/Users/shelly/Documents/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 459, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: slow_conv2d_forward_mps: input(device='cpu') and weight(device=mps:0')  must be on the same device

**Reply by sonluong1987** [reply]
@shelly-luo I believe u need to add this code to body.py line 48 

    if torch.backends.mps.is_available():
        mps_device = torch.device("mps:0")
        data = data.to(mps_device)

**Comment by AlanZhou2022** [comment]
Hi, my problem is that controlnet doesn't take effect when processing pictures in my M2 macbook and no mistakes was showed. (i'm sure that I strictly follow the guidelines, stable diffusion V1.5.1 with the newest version of controlnet) I'm not sure if anybody have the same problem with me?

**Reply by JohanBinst** [reply]
I'm having the same issue on a M2. I used a xy plot with different control net models and weights and all the output images we're the same.

**Comment by foxstudio** [comment]
That's great to hear! I'm glad you were able to solve the issue. I hope your experience with the Mac Studio M1 Ultra chip has been helpful to everyone. If you have any other questions or need further assistance, feel free to ask.

File: `../comfyui/custom_node/facerestore/facelib/detection/retinaface/retinaface.py`
<img width="857" alt="Snipaste_22-24-34" src="https://github.com/Mikubill/sd-webui-controlnet/assets/3861843/2c64e7ea-d772-403f-abfa-973cc37f730e">

**Comment by Lichtfabrik** [comment]
Hi there, followed the instruction to fix this issue, but I still got this Error. Only when I want to use AnimateDiff in combination with ControllNet on A1111. 

RuntimeError: slow_conv2d_forward_mps: input(device='cpu') and weight(device=mps:0')  must be on the same device

I have added the lines in the body.py etc. Any idea what I could do else?

**Comment by AlanZhou2022** [comment]
Many errors occurred in Mac, I think SD is not compatible with Mac very well, but it works well in laptop with Nvidia.


---------------- 

----------原邮件信息----------
发件人："JohanBinst" ***@***.***>
收件人："Mikubill/sd-webui-controlnet" ***@***.***>
抄送人："Alan Zhou" ***@***.***>, "Comment" ***@***.***>
主题：Re: [Mikubill/sd-webui-controlnet] Can't get it to run on Applesilicon. (Discussion #860)
发送时间：2024-02-27 14:48:47

 I'm having the same issue on a M2. I used a xy plot with different control net models and weights and all the output images we're the same.

--
Reply to this email directly or view it on GitHub:
https://github.com/Mikubill/sd-webui-controlnet/discussions/860#discussioncomment-8601304
